Issue #2799361 - Port condition expression action links to the UI.

dsasser opened the pull request:
Adds the ability to group conditions within AND and OR condition sets in the UI. I used the patch from #452 to provide drag and drop functionality to the form and modified from there. See the d.o issue at: https://www.drupal.org/node/2799361

Commit:
#2799361 - Add drag and drop to actions and conditions from #2648334.

Commit:
#2799361 - Create  rules_and and rules_or condition sets (expressions) in the UI.

kevin-dutra commented on src/Form/Expression/ActionContainerForm.php:
```diff
@@ -34,36 +34,80 @@ public function __construct(ActionExpressionContainerInterface $action_set) {
+      $rules_ui_handler = $this->getRulesUiHandler();
+      $row['operations'] = [
+        'data' => [
+          '#type' => 'dropbutton',
```
Technically not part of the new changes, but there's a dedicated `Operations` render element that extends from `Dropbutton`.

dsasser replied:
The dropdown was part of drag/drop PR #452 that this PR is using. But knowing that the `operations` element adds a theme suggestion is a good reason to change it.

kevin-dutra commented on src/Form/Expression/ConditionContainerForm.php:
```diff
@@ -24,65 +25,356 @@ class ConditionContainerForm implements ExpressionFormInterface {
+   */
+  private function mirrorElements($values) {
+    $elements = NULL;
+    $expressionManager = \Drupal::service('plugin.manager.rules_expression');
```
Ideally this service would be introduced via dependency injection, rather than doing the global call. As an example, see `ContentEntityForm`.

dsasser replied:
I agree that dependency injection would be ideal. The reason its not implemented with DI is because the form is being called in a non-standard way (not from the container). Its getting a single argument passed into the constructor, and so we would have to go further up into the api to affect that. I wanted to stay away from making api changes as much as possible.

kevin-dutra replied:
Ah, okay. Maybe abstract it out into a separate method, e.g.

```
$expressionManager = $this->getExpressionManger();
...
function getExpressionManger() {
  return \Drupal::service('plugin.manager.rules_expression');
}
```

Mainly I'm just trying to keep options open for unit testing.

kevin-dutra commented on src/Plugin/RulesExpression/ActionSet.php:
```diff
@@ -27,7 +27,17 @@ protected function allowsMetadataAssertions() {
   public function executeWithState(ExecutionStateInterface $state) {
+    // Get hold of actions.
+    // @todo See if we can add getExpressions method of ExpressionContainerBase.
+    $actions = [];
```
Is it really necessary to put in an extra loop just for the copying to a local variable? Can't you just
`$actions = $this->actions;` or even just sort the class property directly? (Ditto for the other applicable classes.)

dsasser replied:
Actually I think its necessary because $this->actions is really an iterable object and not an array. This is part of the code coming from the included PR #452.

kevin-dutra replied:
Hmm...looks like an array to me:

```
  /**
   * List of actions that will be executed.
   *
   * @var \Drupal\rules\Engine\ActionExpressionInterface[]
   */
  protected $actions = [];

  ...

  public function __construct(array $configuration, $plugin_id, $plugin_definition, ExpressionManagerInterface $expression_manager) {

    ...

    foreach ($configuration['actions'] as $action_config) {
      $action = $expression_manager->createInstance($action_config['id'], $action_config);
      $this->actions[] = $action;
    }
  }
```

But at any rate, if it's part of the other PR, then I won't split hairs here.

kevin-dutra commented on src/Form/Expression/ConditionContainerForm.php:
```diff
@@ -24,65 +25,356 @@ class ConditionContainerForm implements ExpressionFormInterface {
+              '@plugin' => $this->conditionContainer->getLabel(),
+              '@config-plugin' => $config->bundle(),
+              '%label' => $config->label(),
+              '@url' => 'http://drupal.org/node/1300034',
```
There is a newer placeholder type that should be used here, I think: https://www.drupal.org/node/2571689

dsasser replied:
Good to know about the newer placeholder. I assume we are talking about the @url element? Since its not user provided input I wouldn't normally worry about XSS escaping for a hard coded url, but I can change it to use the new placeholder quick enough that its a moot point.

kevin-dutra replied:
Yep, `@url`.

kevin-dutra reviewed (changes requested):
A few minor things that I can see.

Commit:
WDC-16595 Updates based on code review;
- Replaced dropbutton with operation elements
- Used new :placeholder for url placeholder values
- Move expressionManager to class property

kevin-dutra reviewed (approved):
Cool beans.

Commit:
Issue #2799361 by coderwan: Ensure component config entity is updated from component changes during save.
